File: composição.java
import java.util.ArrayList;

class Motor {
    private String tipo;
    private int potencia;

    public Motor(String tipo, int potencia) {
        this.tipo = tipo;
        this.potencia = potencia;
    }

    public void ligar() {
        System.out.println("O motor está ligado");
    }

    public void desligar() {
        System.out.println("O carro está desligado");
    }
}

class Pneu {
    private String marca;
    private int tamanho;

    public Pneu(String marca, int tamanho) {
        this.marca = marca;
        this.tamanho = tamanho;
    }

    public void inflar() {
        System.out.println("O pneu está inflado");
    }

    public void desinflar() {
        System.out.println("O pneu está desinflado");
    }
}

class Carro {
    private String marca;
    private String modelo;
    private Motor motor;
    private ArrayList<Pneu> pneus;

    public Carro(String marca, String modelo) {
        this.marca = marca;
        this.modelo = modelo;
        this.motor = new Motor("Gasolina", 150);
        this.pneus = new ArrayList<>();
        for (int i = 0; i < 4; i++) {
            pneus.add(new Pneu("Pirelli", 18));
        }
    }

    public void ligar() {
        motor.ligar();
        System.out.println("O carro está pronto para rodar");
    }

    public void desligar() {
        motor.desligar();
        System.out.println("O carro foi desligado");
    }

    public void trocarPneu(int indice, Pneu novoPneu) {
        pneus.set(indice, novoPneu);
        System.out.println("Pneu " + (indice + 1) + " trocado");
    }
}

public class composição {
    public static void main(String[] args) {
        Carro carro = new Carro("Toyota", "Corolla");
        carro.ligar();

        Pneu novoPneu = new Pneu("Michelin", 1);
        carro.trocarPneu(2, novoPneu);
        carro.desligar();
    }
}

File: composição.go
package main

import "fmt"

type Motor struct {
	tipo     string
	potencia int
}

func (m Motor) Ligar() {
	fmt.Println("O motor está ligado")
}

func (m Motor) Desligar() {
	fmt.Println("O carro está desligado")
}

type Pneu struct {
	marca   string
	tamanho int
}

func (p Pneu) Inflar() {
	fmt.Println("O pneu está inflado")
}

func (p Pneu) Desinflar() {
	fmt.Println("O pneu está desinflado")
}

type Carro struct {
	marca  string
	modelo string
	motor  Motor
	pneus  []Pneu
}

func NovoCarro(marca string, modelo string) Carro {
	motor := Motor{tipo: "Gasolina", potencia: 150}
	pneus := make([]Pneu, 4)
	for i := 0; i < 4; i++ {
		pneus[i] = Pneu{marca: "Pirelli", tamanho: 18}
	}
	return Carro{marca: marca, modelo: modelo, motor: motor, pneus: pneus}
}

func (c Carro) Ligar() {
	c.motor.Ligar()
	fmt.Println("O carro está pronto para rodar")
}

func (c Carro) Desligar() {
	c.motor.Desligar()
	fmt.Println("O carro foi desligado")
}

func (c *Carro) TrocarPneu(indice int, novoPneu Pneu) {
	c.pneus[indice] = novoPneu
	fmt.Printf("Pneu %d trocado\n", indice+1)
}

func composição() {
	carro := NovoCarro("Toyota", "Corolla")
	carro.Ligar()

	novoPneu := Pneu{marca: "Michelin", tamanho: 1}
	carro.TrocarPneu(2, novoPneu)
	carro.Desligar()
}
